pyriemann/spatialfilters.py

```diff
@@ -462,3 +464,312 @@ def fit(self, X, y):
+        processes, the first cospectrum (0 Hz) must be excluded.
+    fmax : float | None , (default None)
+        The maximal frequency to be returned.
+    fs : float | None , (default None)
+        The sampling frequency of the signal.
+    expl_var : float (default 0.999)
+        The percentage of explained variance in ]0, 1], for dimension reduction
+        of cospectra, because Pham's AJD is sensitive to matrices conditioning.
+    verbose : bool (default True)
+        Verbose flag.
+
+    Attributes
+    ----------
+    n_channels_ : int
+        If fit, the number of channels of the signal.
+    freqs_ : ndarray , shape (number of frequencies,)
+        If fit, the frequencies associated to cospectra.
+    n_sources_ : int
+        If fit, the number of components of the source space.
+    forward_filters_ : ndarray , shape ``(n_sources_, n_channels_)``
+        If fit, the spatial filters used to transform signal into source,
+        also called deximing or separating matrix.
+    backward_filters_ : ndarray , shape ``(n_channels_, n_sources_)``
+        If fit, the spatial filters used to transform source into signal,
+        also called mixing matrix.
+
+    Notes
+    -----
+    .. versionadded:: 0.2.7
+
+    See Also
+    --------
+    CospCovariances
+
+    References
+    ----------
+    [1] M. Congedo, C. Gouy-Pailler, C. Jutten, "On the blind source separation
+    of human electroencephalogram by approximate joint diagonalization of
+    second order statistics", Clin Neurophysiol, 2008
+
+    [2] M. Congedo, R. John, D. De Ridder, L. Prichep, "Group indepedent
+    component analysis of resting state EEG in large normative samples",
+    Int J Psychophysiol, 2010
+
+    [3] D.-T. Pham, "Joint approximate diagonalization of positive definite
+    Hermitian matrices", SIAM J Matrix Anal Appl, 2001
+    """
+
+    def __init__(self, window=128, overlap=0.5, fmin=None, fmax=None, fs=None,
+                 expl_var=0.999, verbose=True):
+        """Init."""
+        if not 0 < expl_var <= 1:
+            raise ValueError('Parameter expl_var must be included in ]0, 1]')
+
+        self.window = window
+        self.overlap = overlap
+        self.fmin = fmin
+        self.fmax = fmax
+        self.fs = fs
+        self.expl_var = expl_var
+        self.verbose = verbose
+        #TODO: add an argument 'algo' to choose AJD: pham2001 or ablin2019
+
+    def fit(self, X, y=None):
+        """Fit.
+
+        Compute and diagonalize cospectra, to estimate forward and backward
+        spatial filters.
+
+        Parameters
+        ----------
+        X : ndarray, shape (n_subjects, n_conditions, n_channels, n_samples) | list of n_subjects of list of n_conditions ndarray of shape (n_channels, n_samples), with same n_conditions and n_channels but different n_samples
+            ndarray of signal in channel space, acquired for different subjects
+            and under different experimental conditions.
```

Comment:
fit and transform have different input. 
`(n_trials, n_channels, n_samples) # in transform`

should it be better to have also `X shape (n_trials, n_channels, n_samples)` and use `y` to find the conditions ?

Reply:
Thanks, it's a good idea! 

But, what could be the optimal shape and content of `y`? 

It could be a ndarray of shape `(n_subjects, n_conditions, 3)`, where the final dimension contains 3 values: 1 condition index, and 2 sample indices, defining the beginning and the end of the condition.

Reply:
More I think about this and more it seems not trivial (so not urgent).

The only default I see have having X different in fit/transform is that it can be confusing for the user but the proposal bellow is even more confusing. So keep it with a grain of salt just for discussing the design, and your solution seems good enough.


- if `X` is `(n_trials, n_channels, n_samples)` 3D epochs. Note that most of the time it will be 2D as users just want to find the sources of the full EEG signal, i.e.  `(n_channels, n_samples)`.
- then `y`, taking inspiration of the sklearn API would be either (n_trials,) with n_classes unique labels or (n_trials, n_classes).

So we would have the following options for `y`
```
- None  # (no condition, no specific subject) : simple BSS or gBSS scheme of AJD where `n_trials` cross-spectral matrices are averaged. So user would have to unsure herself conditions and subjects balance, i.e. giving the same number of trial for a given subject or condition. Easy to go. Should work for most situation.
- ndarray, shape (n_trials,)  # supervised training n_condition*n_subjects unique labels for AJD of cross-spectral matrices. SCM are averaged for each unique label (condition or subject or subject+condition)*
- ndarray, shape (n_trials, n_classes)  # same than before but where n_classes =  n_condition*n_subjects and boolean instead
```

*e.g. if you want to average subjects for gBSS, you could add an input "labels_to_average" which can be a list or list of list, e.g.
```
labels_to_average :  list | list of list (default: None)
    List of unique labels in `y` which will be averaged together.
    By default (None), each unique labels has one cross-spectrum matrix estimation.
    Example for group Blind Source Separation (gBSS) with uniform cohort:
    >>> labels_to_average = ['subject_1', 'subject_2', ...., 'subject_M']
    Example for group Blind Source Separation (gBSS) with non-uniform cohort:
    >>> labels_to_average = [['subject_1', 'subject_2'],
    >>>                                   ['subject_3', 'subject_4']]  # subjects 1&2 belong to different cohort than 3&4

    Example for grouping conditions together
    >>> labels_to_average = [['condition_1', 'condition_2'],
    >>>                                    ['condition_3']]  # two first conditions are merged together

``` 

I should ask @Marco-Congedo because IMO, averaging CSM for all subjects may not be the only use of gBSS as sources similarities may be just as important as sources dissimilarities. Therefore I wonder if BSS could use be used for different cohorts or a completely blind scheme where each subject has his own Cross-Spectral Matrix.

Reply:
For `transform` and `inverse_transform` functions, `trials` can relate to any batch of epochs, coming from different subjects, sessions, conditions, trials, etc.

For the `fit` function, the name `trials` seems confusing: we don't know if this quantity relates to `subjects` or `conditions`. It is a crucial problem because these dimensions are not interchangeable in `AJDC`. They are not processed in the same way: cospectra are averaged across subjects but concatenated along conditions. So, the shape of `y` must make it possible to contain these two dimensions and distinguish them. Adding an argument `labels_to_average` breaks the compatibility with `sklearn` API.

Unfortunately, we can't mimic `sklearn` or `mne` APIs, which provide functions for BSS (notably ICA) but not for group BSS (pyRiemann is going to fast!).

**Suggestion for `fit` arguments**
```
X : ndarray, shape (n_trials, n_channels, n_samples)
    ndarray of trials in channel space
y : list of strings of shape (n_trials,) | None
    List of labels, characterizing each trial with respect its subject and/or condition (semicolon separated)
```

Example of `y` for:
BSS under a single condition : `None`
BSS under several conditions : `['condition_baseline', 'condition1', 'condition_2']`
gBSS under a single condition : `['subject1', ..., 'subject_17', ..., 'subject-42bis']`
 gBSS under several conditions : `['subject1;condition1', 'subject_2;condition1', 'subject3;condition-2', 'subject3;condition1', 'subject1;condition-2', 'subject_2;condition-2']`

With this specification, `y` could be decoded in two steps:
- using regex on `subject` and `condition` expressions, to know what the label refers to,
- and then counting the number of different labels for each of the two expressions.

Then, we could reshape trials X with respect found labels. Then AJDC could begin...
It makes the job, but it still seems cumbersome compared to the current implementation.

Reply:
indeed. Let's keep your implementation.